cmd/libsnap-confine-private/cgroup-support.c

```diff
@@ -97,3 +101,49 @@ bool sc_cgroup_is_v2() {
     }
     return false;
 }
+
+static void ensure_dir(const char *dir, mode_t mode) {
+    struct stat stat_buf;
+
+    /* The path /run/snapd should be a directory with mode 0655 owned by
+     * root.root. If one is absent it is created. If one is there it is
+     * validated for correct type. */
+    if (lstat(dir, &stat_buf) < 0) {
+        if (errno != ENOENT) {
+            die("cannot lstat %s", dir);
+        }
+        if (mkdir(dir, mode) < 0) {
+            die("cannot mkdir %s", dir);
+        }
+        if (chown(dir, 0, 0) < 0) {
+            die("cannot chown %s to root.root", dir);
+        }
+    } else {
+        if ((stat_buf.st_mode & S_IFMT) != S_IFDIR) {
+            die("cannot proceed: %s must be a directory", dir);
+        }
+    }
+}
+
+void sc_cgroup_mount_snapd_hierarchy(void) {
+    ensure_dir(snapd_run_dir, 0755);
+    ensure_dir(snapd_run_cgroup_dir, 0755);
+
+    /* The path /run/snapd/cgroup should be a mount point for a cgroup. */
+    struct statfs statfs_buf;
+    if (statfs(snapd_run_cgroup_dir, &statfs_buf) < 0) {
+        die("cannot statfs %s", snapd_run_cgroup_dir);
+    }
+    if (statfs_buf.f_type != CGROUP_SUPER_MAGIC) {
+        int mount_flags = MS_NOSUID | MS_NODEV | MS_NOEXEC | MS_RELATIME;
+        // Here "none" indicates that no controllers are enabled.
+        const char *mount_opts = "none,name=snapd";
+        if (mount("cgroup", snapd_run_cgroup_dir, "cgroup", mount_flags, mount_opts) < 0) {
```

Comment:
`man cgroups` gives the example of `mount -t cgroup -o none,name=somename none /some/mount/point` and  stracing mount has it use the "none" as the source as well. Perhaps for consistency with mount, use: `if (mount("none", snapd_run_cgroup_dir, ...`

Reply:
I chose to be consistent with how systemd mounts cgroups. See https://github.com/systemd/systemd/blob/86e94d95d020c34403bc99675318ed4e609ea3a0/src/core/mount-setup.c#L304 for details.